bd-client-stats/src/file_manager.rs

```diff
@@ -0,0 +1,424 @@
+        &compressed,
+      )
+      .await?;
+
+    Ok(())
+  }
+
+  async fn delete_oldest_snapshot(&mut self) -> anyhow::Result<()> {
+    self
+      .file_system
+      .delete_file(&STATS_DIRECTORY.join(&self.index[0].name))
+      .await?;
+    self.index.pop_front();
+
+    Ok(())
+  }
+
+  async fn delete_pending_upload(&mut self) -> anyhow::Result<()> {
+    log::debug!("deleting pending upload: {}", self.index[0].name);
+    self.delete_oldest_snapshot().await?;
+    self.write_index().await?;
+
+    Ok(())
+  }
+}
+
+impl Inner {
+  // Initialize or get an already initialized file manager.
+  async fn get_initialized(&mut self) -> anyhow::Result<&mut InitializedInner> {
+    // Due to the way logger startup works we delay initializing the index until first use. First we
+    // try to load an existing file index. If this doesn't exist or is corrupted we remove the
+    // entire directory as there is no reasonable way to manage the contents.
+    match self {
+      Self::Initialized(inner) => Ok(inner),
+      Self::NotInitialized(file_system) => {
+        let file_system = file_system.take().unwrap();
+        let path = STATS_DIRECTORY.join(&*PENDING_AGGREGATION_INDEX_FILE);
+        log::debug!("initializing pending aggregation index: {}", path.display());
+        let index = match file_system
+          .read_file(&path)
+          .await
+          .and_then(|contents| read_compressed_protobuf(&contents))
+        {
+          Ok(index) => index,
+          Err(e) => {
+            log::debug!("unable to open pending aggregation index: {e}");
+            log::debug!("creating new aggregation index");
+
+            file_system.remove_dir(&STATS_DIRECTORY).await?;
+            file_system.create_dir(&STATS_DIRECTORY).await?;
+            PendingAggregationIndex::default()
+          },
+        };
+
+        *self = Self::Initialized(InitializedInner {
+          file_system,
+          index: index.pending_files.into(),
+        });
+
+        Ok(match self {
+          Self::Initialized(inner) => inner,
+          Self::NotInitialized(_) => unreachable!(),
+        })
+      },
+    }
+  }
+}
+
+impl FileManager {
+  pub fn new(
+    file_system: Box<dyn FileSystem>,
+    time_provider: Arc<dyn TimeProvider>,
+    runtime_loader: &ConfigLoader,
+  ) -> anyhow::Result<Self> {
+    Ok(Self {
+      inner: Mutex::new(Inner::NotInitialized(Some(file_system))),
+      time_provider,
+      max_aggregated_files: runtime_loader.register_watch()?,
+      max_aggregation_window_per_file: runtime_loader.register_watch()?,
+    })
+  }
+
+  // Read an existing snapshot from disk to merge into, or create a new one.
+  pub async fn get_or_create_snapshot(&self) -> anyhow::Result<StatsUploadRequestHandle> {
+    let mut inner = self.inner.lock().await;
+    let initialized_inner = inner.get_initialized().await?;
+
+    let use_existing = if initialized_inner.index.back().map_or(true, |file| {
```

Comment:
This block is a bit hard to read, maybe pull the "create new snapshot check" out into a var instead of this big if condition?

Reply:
Will do.